tools/autograd/gen_autograd_functions.py

```diff
@@ -63,6 +63,12 @@
 }
 """)
 
+DERIVATIVE_MULTI_COPY_RANGE = CodeTemplate("""\
+  if (should_compute_output({ ${name}_ix })) {
+    copy_range(grad_inputs, ${name}_ix, std::get<${i}>(grad_result));
+  }
```

Comment:
Wouldn't this cause some of the returned gradients to be undefined? I think we have a pretty strict no-undefined policy in autograd, so we might still allocate zero inputs. Am I right @colesbury?

Reply:
From inspecting the generated Functions.cpp I had the impression that the end result when should_compute_output now is exactly the same as it were when the other grads of the "MULTI" had should_compute_output.
I must admit that I still cannot figure out what to change when revisiting Functions.cpp with your comment in mind.

Reply:
Oh, the above is hard to read without the context of the actual code. The generated snippet for cudnn rnn backward *before* the PR is:
```
  if (should_compute_output({ input_ix, hx_ix, cx_ix, weight_ix })) {
      auto grad_input_mask = std::array<bool, 4>{
      should_compute_output({ input_ix }),
      should_compute_output({ hx_ix }),
      should_compute_output({ cx_ix }),
      should_compute_output({ weight_ix }),
    };
    auto grad_result = _cudnn_rnn_backward(input, weight, weight_stride0, result4, hx, cx, result0, grads[0], grads[1], grads[2], mode, hidden_size, num_layers, $
    copy_range(grad_inputs, input_ix, std::get<0>(grad_result));
    copy_range(grad_inputs, hx_ix, std::get<1>(grad_result));
    copy_range(grad_inputs, cx_ix, std::get<2>(grad_result));
    copy_range(grad_inputs, weight_ix, std::get<3>(grad_result));
  }
```
and afterwards, you get the additional `if`s around the copy_range calls.
Looking at that, I might use `grad_input_mask` instead of `should_compute_output` if you prefer.

Reply:
Yes, this looks right to me. If we are not requested to compute output for something, we're allowed to return undefined.